jax/_src/clusters/cluster.py

```diff
@@ -0,0 +1,78 @@
+from typing import List
+from absl import logging
+from jax._src.cloud_tpu_init import running_in_cloud_tpu_vm
+
+
+class ClusterEnv:
+  """Interface for defining a cluster environment.
+
+  To enable auto bootrapping (aka :func:`jax.distributed.initialize()`),
+  cluster environments need to derive from :class:`ClusterEnv` and implement
+  :func:`is_env_present`, :func:`get_coordinator_address`,
+  :func:`get_process_count`, and :func:`get_process_id`.
+  """
+
+  _cluster_types: List['ClusterEnv'] = []
+
+  def __init_subclass__(cls, **kwargs):
+    super().__init_subclass__(**kwargs)
+    cls._cluster_types.append(cls)
+
+  @classmethod
+  def auto_detect_unset_distributed_params(cls, coordinator_address,
+                                           num_processes, process_id,
+                                           local_device_ids):
+    if all(p is not None for p in (coordinator_address, num_processes,
+      process_id, local_device_ids)):
+      return (coordinator_address, num_processes, process_id,
+              local_device_ids)
+    env = next((env for env in cls._cluster_types if env.is_env_present()), None)
+    if env:
+      logging.info('Auto-bootstrapping JAX processes via %s', env.__name__)
+      if coordinator_address is None:
+        coordinator_address = env.get_coordinator_address()
+      if num_processes is None:
+        num_processes = env.get_process_count()
+      if process_id is None:
+        process_id = env.get_process_id()
+      # Never automatically set local_device_ids on TPUs
+      # Default to single process per device otherwise
+      if local_device_ids is None and not running_in_cloud_tpu_vm:
+        local_device_ids = [env.get_local_process_id()]
+    else:
+      logging.info('Could not find an environment to use for auto-bootstrapping. '
+        'Known environments: %s', ', '.join(e.__name__ for e in cls._cluster_types))
+    return (coordinator_address, num_processes, process_id,
+            local_device_ids)
+
+  @classmethod
+  def is_env_present(cls):
+    """Return True if process is running in this cluster environment.
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def get_coordinator_address(cls):
+    """Get address and port used by JAX to bootstrap.
+
+    Process id 0 will open a tcp socket at "hostname:port" where
+    all the proccesses will connect for bootstrapping.
+    The selected port needs to be free.
+    :func:`get_coordinator_address` needs to return the same hostname and port on all the processes.
+
+    Returns:
+      "hostname:port" used for bootstrapping
+    """
+    raise NotImplementedError
+
+  @classmethod
+  def get_process_count(cls):
+    raise NotImplementedError
+
+  @classmethod
+  def get_process_id(cls):
+    raise NotImplementedError
+
+  @classmethod
+  def get_local_process_id(cls):
```

Comment:
This can apparently return None, so I think it needs a docstring explaining what that means. And should this return None by default instead of raising NotImplemented?

Reply:
It cannot be None. To make sure it is dealt with by the subclass, it would let the exception. I can add a docstring on what it means.